storage/innobase/include/clone0clone.h

```diff
@@ -0,0 +1,1360 @@
+        err = func(false, wait);
+        continue;
+      }
+
+      auto alert = (alert_count > 0) ? (loop_count % alert_count == 0) : true;
+
+      err = func(alert, wait);
+
+      is_timeout = (loop_count > total_count);
+    }
+    return (err);
+  }
+
+  /** Wait till the condition is satisfied or default timeout.
+  @param[in]    func            callback function for condition check
+  @param[in]    mutex           release during sleep and re-acquire
+  @param[out]   is_timeout      true if timeout
+  @return error code returned by callback function. */
+  static int wait_default(Wait_Cond_Cbk_Func &&func, mysql_mutex_t *mutex,
+                          bool &is_timeout) {
+    return (wait(CLONE_DEF_SLEEP, Clone_Sec(CLONE_DEF_TIMEOUT),
+                 CLONE_DEF_ALERT_INTERVAL,
+                 std::forward<Wait_Cond_Cbk_Func>(func), mutex, is_timeout));
+  }
+
+  /** Check if any active clone is running.
+  @param[in]    print_alert     print alert message
+  @return true, if concurrent clone in progress */
+  bool check_active_clone(bool print_alert);
+
+  /** Check if any active clone is running.
+  @return (true, handle) if concurrent clone in progress */
+  std::tuple<bool, Clone_Handle *> check_active_clone();
+
+  /** @return GTID persistor */
+  // Clone_persist_gtid &get_gtid_persistor() { return (m_gtid_persister); }
+
+  /** Remember that all innodb spaces are initialized after last startup. */
+  void set_space_initialized() { m_space_initialized.store(true); }
+
+  /** @return true if all innodb spaces are initialized. */
+  bool is_space_initialized() const { return m_space_initialized.load(); }
+
+ private:
+  /** Find free index to allocate new clone handle.
+  @param[in]    hdl_type        clone handle type
+  @param[out]   free_index      free index in array
+  @return error code */
+  int find_free_index(Clone_Handle_Type hdl_type, uint &free_index);
+
+  /** Handle restricted state during critical ddl phase.
+  @param[in]    type    ddl notification type
+  @param[in]    space   tablespace ID for which notification is sent
+  @param[in]    begin   true, if beginning state
+                        false, if ending
+  @return true iff clone needs to wait for state change. */
+  bool handle_ddl_state(Clone_notify::Type type, space_id_t space, bool begin);
+
+ private:
+  /** Array of clone handles */
+  Clone_Handle *m_clone_arr[CLONE_ARR_SIZE];
+
+  /** Number of copy clones */
+  uint m_num_clones;
+
+  /** Number of apply clones */
+  uint m_num_apply_clones;
+
+  /** Array of clone snapshots */
+  Clone_Snapshot *m_snapshot_arr[SNAPSHOT_ARR_SIZE];
+
+  /** Number of copy snapshots */
+  uint m_num_snapshots;
+
+  /** Number of apply snapshots */
+  uint m_num_apply_snapshots;
+
+  /** Clone system mutex */
+  mysql_mutex_t m_clone_sys_mutex;
+
+  /** Clone unique ID generator */
+  uint64_t m_clone_id_generator;
+
+  /** If all innodb tablespaces are initialized. */
+  std::atomic<bool> m_space_initialized;
```

Comment:
It is unclear why this member is needed. Can `innodb_clone_begin()` be invoked several times, "incrementally"?